### PR TITLE
EES-1730 Add initial performance optimizations for charts/tables

### PR DESCRIPTION
**Pull request by ntsim**
This PR adds the following performance optimizations:

- Adds pre-grouping of result measures by their respective set of filters. This drastically speeds up lookup of results during `createDataSetCategories` (up to ~10x faster)
- Adds a similar pre-grouping of result measures optimization to `TimePeriodDataTable` as well (~4-5x faster)
- Aggregates multiple operations required for calculating `MapBlock` geometries into a single operation (~0.25x faster)
- Removes client-side camelCasing of `geographicLevel` for both charts and tables. This only shaves off a fairly neglible amount of time (~1%), but it mainly makes sense to push any camelCasing onto the server for consistency.

## Relevant changes

- Fixes `ReleaseDataBlockSourceWizardFinalStep` rendering its table twice unnecessarily.
- Changes 'Table' tab in `ReleaseDataBlocksPageTabs` to be rendered lazily i.e. will only be rendered when the user visits the tab. This should help mitigate potentially slow initial renders.

## Benchmark notes

I benchmarked by creating two data blocks from absence data. They were created with map charts using the following filters:

- Block 1: all local authorities, 5 time periods, 3 indicators  and 1 filter
- Block 2: all local authorities, 1 time period, 3 indicators and 2 school types

### Baseline

- Block 1
  - Map: 2566.8ms
  - Table: 428.6ms
- Block 2
  - Map: 174ms
  - Table: 174ms

### After optimizations

- Block 1
  - Map: 199.9 - 203.3ms
  - Table: 91.5ms
- Block 2
  - Map: 101.2 - 114.8ms
  - Table: 43ms

## Further work

There are most likely some more minor performance tweaks that we can do to the main bottlenecked areas (`createDataSetCategories` and `TimePeriodDataTable`), however, these may not have a very good benefit/effort ratio. These may only yield 10-20% speed ups for quite a lot of refactoring and general decrease in readability.

The crux of the issue is that for large tables/charts, there are just a lot of iterations that have to be performed as there is a lot of data to crunch! Additionally, a fairly large chunk of execution time is out of our control as it's up to Leaflet to actually render the maps.

A better investment of time would be to try and remove/reduce UI freezes so that users can't really perceive any slowness. Ideally we want to offload any processing to another thread before the table/chart renders. Web workers will most likely be needed for this (see [EES-1809](https://dfedigital.atlassian.net/browse/EES-1809)).

### CODE REVIEW

File: src/explore-education-statistics-common/src/modules/table-tool/components/TimePeriodDataTable.tsx [hunk 1]
```diff
@@ -99,39 +152,41 @@ const TimePeriodDataTableInternal = forwardRef<HTMLElement, Props>(
               ...columnFilterCombination,
             ];
 
-            const matchingResult = results.find(result => {
-              return filterCombination.every(filter => {
+            const dataSet = filterCombination.reduce<
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-common/src/modules/table-tool/components/TimePeriodDataTable.tsx:155]: Ahh this looks very promising!

File: src/explore-education-statistics-common/src/modules/table-tool/components/__tests__/TimePeriodDataTable.test.tsx [hunk 2]
```diff
@@ -264,8 +266,8 @@ describe('TimePeriodDataTable', () => {
       results: [
         {
           filters: ['598ed9fd-b37e-4e08-baec-08d78f6f2c4d'],
-          geographicLevel: 'Country',
-          location: { country: { code: 'E92000001', name: 'England' } },
+          geographicLevel: 'localAuthority',
```
Comment (duncan-at-hiveit) [inline src/explore-education-statistics-common/src/modules/table-tool/components/__tests__/TimePeriodDataTable.test.tsx:269]: Why the change from "country" type to "localAuthority" here?  Is it just to make the grouping functions' test setups easier?

**Review by duncan-at-hiveit — APPROVED** [review]
Just left some questions if you could take a look?  Ta.

File: src/GovUk.Education.ExploreEducationStatistics.Data.Services/ViewModels/ObservationViewModel.cs [hunk 3]
```diff
@@ -12,13 +10,13 @@ public class ObservationViewModel
         public IEnumerable<string> Filters { get; set; }
 
         // TODO could this be [JsonConverter(typeof(EnumToEnumValueJsonConverter<GeographicLevel>))]?
-        [JsonConverter(typeof(StringEnumConverter))]
+        [JsonConverter(typeof(StringEnumConverter), true)]
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/GovUk.Education.ExploreEducationStatistics.Data.Services/ViewModels/ObservationViewModel.cs:13]: This enables camelCasing :smile:

File: src/explore-education-statistics-common/src/modules/table-tool/components/DownloadCsvButton.tsx [hunk 4]
```diff
@@ -44,7 +43,7 @@ export const getCsvData = (fullTable: FullTable): string[][] => {
 
       const indicatorCells = indicators.map(indicator => {
         const matchingResult = results.find(result => {
-          const geographicLevel = camelCase(result.geographicLevel);
+          const { geographicLevel } = result;
```
[the remark this answers is not in the document]
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/table-tool/components/DownloadCsvButton.tsx:46]: Yep indeed, we no longer need to do this on the client, speeding things up a little bit.

File: src/explore-education-statistics-common/src/modules/table-tool/components/__tests__/TimePeriodDataTable.test.tsx (hunk 2 above)
Comment (ntsim) [reply src/explore-education-statistics-common/src/modules/table-tool/components/__tests__/TimePeriodDataTable.test.tsx:269]: Find and replace gone wrong I'm afraid - fixed!
